Better pruning of events from Pythia8 and enable in HepMC

cholmcc opened the pull request:
This commit
- Fixes the pruning of events in `GeneratorPythia8`.  The old implementation had a number of problems

  - Beam particles (HepMC status = 4) was _not_ kept
  - The event structure was _not_ preserved.  That meant that the event would be split into many disjoint trees.  Thus, subsequent passes over the event would not be able to tie things together.  This was especially a problem when converting the event to HepMC (via AOD) since the events would not be complete.
- Added the option to prune events read in via `GeneratorHepMC`.  The configuration key `HepMC.prune=true` will enable this mode (off by default).

By _pruning_, I mean that all particles that are _not_
- beam particles (HepMC status = 4),
- decayed (HepMC status = 2), nor
- final state (HepMC status = 1) are removed from the event.  When we do that, it is extremely important to connect up particles so that we have a consistent event structure. For example, we may remove a gluon from the tree.  In that case, we must take care to move all child particles to the proper node in the tree, or the children will be disjoint from the rest of the event.

This patch also adds Doxygen documentation to `GeneratorPythia8`, as well as a new `run/SimExamples/Pythia8/README.md` file.

cholmcc commented:
As an illustration of the problem I attach a PDF.  Each page is the same event shown in 3 different cases 
- `full` is the full event 
- `Pruned (old)` is the event pruned according to the old algorithm 
- `Pruned (new)` is the event pruned using the new algorithm 

I use the script `hepmc.py` to illustrate the event.  That is, I take the event as formatted by HepMC (`HepMC3::Pythia8ToHepMC3`).   Of course, `GeneratorPythia8` does not use that intermediate format, and the particles pushed the simulation stack is different from what is shown.  However, the plots do illustrate the problem - that the events are really messed up. 
[prune.pdf](https://github.com/AliceO2Group/AliceO2/files/13400557/prune.pdf)

alibuild commented:
Error while checking build/O2/fullCI for 87364a7c17f0d293c8958de8340a8eed9ba964ae at 2023-11-18 16:26:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/aba82e2ee3f1d6fefa4ef6aaf77ab6755299b777/slc8_x86-64/o2checkcode/1.0-local152/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12275/87364a7c17f0d293c8958de8340a8eed9ba964ae/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 68dec33256685a53718f3e4b7195f4eb91d6e5d5 at 2023-11-18 16:41:
```
## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/12275-slc8_x86-64/0/Generators/src/GeneratorPythia8.cxx:408:20: error: statement should be inside braces [readability-braces-around-statements]
/sw/SOURCES/O2/12275-slc8_x86-64/0/Generators/src/GeneratorPythia8.cxx:478:42: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12275/68dec33256685a53718f3e4b7195f4eb91d6e5d5/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 33d1d470ea9c4af747ab2b675962079edef88a72 at 2023-11-20 03:47:
```
## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/a9cf90f74024b7f4676c125e55e84c223a8e2462/slc8_x86-64/o2checkcode/1.0-local172/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12275/33d1d470ea9c4af747ab2b675962079edef88a72/build_O2_fullCI/pretty.html).

cholmcc commented:
The test `build/AliceO2/O2/o2/macOS` seems to run with old code - in the quoted `serverlog` the full Pythia event is listed, which I took out of the code.

cholmcc committed:
This commit

Fixes the pruning of events in GeneratorPythia8. The old implementation had a number of problems

- Beam particles (HepMC status = 4) was not kept
- The event structure was not preserved.

That meant that the event would be split into many disjoint trees. Thus, subsequent passes over the event would not be able to tie things together. This was especially a problem when converting the event to HepMC (via AOD) since the events would not be complete.

Added the option to prune events read in via GeneratorHepMC. The configuration key HepMC.prune=true will enable this mode (off by default).

By pruning, I mean that all particles that are not

- beam particles (HepMC status = 4),
- decayed (HepMC status = 2), nor
- final state (HepMC status = 1)

are removed from the event. When we do that, it is extremely important to connect up particles so that we have a consistent event structure. For example, we may remove a gluon from the tree. In that case, we must take care to move all child particles to the proper node in the tree, or the children will be disjoint from the rest of the event.

This patch also adds Doxygen documentation to GeneratorPythia8, as well as a new run/SimExamples/Pythia8/README.md file.

alibuild commented:
Error while checking build/O2/fullCI for a83d8ef2bc0ec92fba351683947e23f1ab1c451c at 2023-11-20 12:29:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/c4257f6798b58be1e1f4f36247ebbde03f4b92f2/slc8_x86-64/o2checkcode/1.0-local179/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12275/a83d8ef2bc0ec92fba351683947e23f1ab1c451c/build_O2_fullCI/pretty.html).

sawenzel commented:
Improvements in the pruning process are greatly appreciated. But for the moment, event generation for PbPb events hangs or is extremely slow (probably explaining the CI failure in fullCI). It looks like the proposed algorithm here is >100x slower than before.

I ran the following command for testing: `o2-sim -j 1 --noGeant -n 1000 -g pythia8hi --seed 11 -o P8HIS11`

sawenzel commented on Generators/include/Generators/GeneratorPythia8.h:
```diff
@@ -88,27 +166,101 @@ class GeneratorPythia8 : public Generator
+   * particle, and thus modifies the passed event structure.
+   * Calculations are cached. */
+  void investigateRelatives(Pythia8::Event& event,                   // Event
+                            const std::map<size_t, size_t>& old2New, // Map from old to new idx
```
Part of the reason why the pruning became much slower might be the use of std::map<> and std::set<>. Access to these scales like O(log(N)) and I would typically advise the use of unordered_map<> for both of these cases (O(1) access).

cholmcc replied:
See longer reply below

sawenzel commented on Generators/src/GeneratorHepMC.cxx:
```diff
@@ -120,36 +121,134 @@ Bool_t GeneratorHepMC::generateEvent()
 }
 
+/*****************************************************************/
+void GeneratorHepMC::pruneEvent(Select select)
```
I think this is a fantastic idea. I was previously wondering if we couldn't prune Pythia8 events by using HepMC API which seems to make this task easy.

We can use this to verify correctness and also to benchmark the Pythia8 pruning function.

cholmcc replied:
Unfortunately, the HepMC pruning is O(N^2 log N), because of HepMCs use of `std::vector`, see also longer reply below.

cholmcc committed:
Speed up of Pythia8 pruning

The pruning of Pythia8 events have been sped up significantly.

The culprit seemed to be the use of `std::set` which on `insert` is
rather slow - O(log(N)).  An alternative would be to use
`std::unordered_set` which is O(1) on average.   However, a different
approach altogether is used which does not require storing all, but only
min/max of, relatives.

The pruning now is roughly O(N log(N)), just as the old algorithm albeit
overall a bit slower.   However, there is no two ways about it - we can
probably not make it faster without corrupting the events.  Note, the
old implementation always _did_ corrupt the events and there was no easy
fix to make sure it didn't.

At the pull request for this change

    https://github.com/AliceO2Group/AliceO2/pull/12275

there is more information.

I could not speed up the HepMC pruning significantly.  The reason for
that is that `HepMC3::GenEvent` and `HepMC3::GenVertex` internally used
`std::vector` for storage.  This implies that removing particles from an
event is _extremely_ slow.  The pruning of HepMC events is therefore
O(N^2 log(N)) - not very nice.

As long as HepMC internally used
`std::vector`, this will always be the case.  A better option would be
to use `std::forward_list`, since access to incoming or outgoing
particles is rarely random but rather sequential.

I was hoping one could build a new event instead, but that turns out to
be impractical.  Mainly, one cannot detach from parent event, nor deep
copy particles or vertexes in an efficient manner.

Pruning of events reduces the event size by roughly 25% to 50% for the
smallest to largest events, respectively.

cholmcc commented:
Hi Sandro,

> Improvements in the pruning process are greatly appreciated. But for the moment, event generation for PbPb events hangs or is extremely slow (probably explaining the CI failure in fullCI). It looks like the proposed algorithm here is >100x slower than before.

Yes, I saw that.  I did a bit of work on this, and managed to speed up the algorithm quite a bit (see also last commit).  The problem was the use of `std::set` and to some extent `std::map`.  I've taken a new approach with is much more efficient and is roughly O(N log N) (as the old one, but with a larger pre-factor). 

It is not possible to get the new algorithm to be as fast as the old _faulty_ algorithm, as far as I can tell.  Thus, I think we must live with the penalty of longer processing time to ensure that the events are not corrupted. 

I similarly also tried to speed up the HepMC pruning, but to little avail.  Here, the culprit is that `HepMC3::GenVertex` and `HepMC3::GenParticle` internally uses `std::vector` to store pointers.  This makes removing particles of vertexes from the event _extremely_ slow.   A better alternative would be to use `std::forward_list`, which I think would be OK because access to vertex incoming and outgoing particles is rarely random but most often sequential.  The HepMC pruning algorithm is O(N^2 log N) - i.e., not good, but I see no way to improve it. 

The plot below shows the effect and timing of pruning of 1000 pp events at 14 TeV. 
 
![prune](https://github.com/AliceO2Group/AliceO2/assets/7202672/2226b1e8-1509-4ae0-87d7-9bf77cd50764)

Here, 
- `Old` is the old _faulty_ algorithm 
- `HepMC` is the algorithm in `GeneratorHepMC` 
- `Pythia` was the algorithm in `GeneratorPythia8` before the last commit 
- `Pythia(New)` is the improved algorithm in `GeneratorPythia8` as in the last commit 

Note, if I do not call `HepMC3::GenEvent::remove_particle` nor `HepMC3::GenEvent::remove_vertex`, then the HepMC line falls on top of the _Pythia(New)_ line. 

I also attach [flame graphs](https://www.brendangregg.com/flamegraphs.html) of the performance of `HepMC`, `Pythia`, and `Pythia(new)`.  in a ZIP archive: [flamegraphs.zip](https://github.com/AliceO2Group/AliceO2/files/13423660/flamegraphs.zip)

The Flame Graphs are dynamic HTML files. 

BTW, doing 
```
> o2-sim --seed 12345 -n 5 --configKeyValues "Diamond.width[2]=6.;" -g pythia8hi -e TGeant4 
...
[INFO] Simulation process took 2434.44 s
...
```
Yours,
_Christian_

cholmcc committed:
Formatting

cholmcc commented:
Hi all, 

For your information, I made a [MR](https://gitlab.cern.ch/hepmc/HepMC3/-/merge_requests/291) against HepMC3 which changes the internal containers from `std::vector` to `std::list`.  This in general speeds up the HepMC3 code (pruning by a factor 3, reading events goes from $`O(n^{3/2}\log n)`$ to $`O(n\log n)`$ ).  However, the good folks over there are very concerned about the API change - which I think isn't really as big a problem as they make it out to be. 

Yours,
_Christian_
